internal/datasource.go

```diff
@@ -0,0 +1,155 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+	"github.com/hashicorp/terraform-plugin-framework/datasource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/opslevel/opslevel-go/v2025"
+)
+
+var _ datasource.DataSourceWithConfigure = (*TFDataSourceSingle[any, any])(nil)
+
+type TFDataSourceSingle[TData any, TModel any] struct {
+	datasource.DataSource
+	client *opslevel.Client
+
+	Name        string
+	Description string
+	Attributes  map[string]schema.Attribute
+
+	ReadFn  func(ctx context.Context, client *opslevel.Client, identifier string) (TData, error)
+	ToModel func(ctx context.Context, identifier string, data TData) (TModel, error)
+}
+
+func (s *TFDataSourceSingle[TData, TModel]) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
```

Comment:
right now this is copy pasted - but we could pull this out because it can be shared between the 2.  I'll wait till i implement the "TFResource` abstraction